docs/dev/sql-nested-function-select-clause.md

```diff
@@ -0,0 +1,476 @@
+
+```mermaid
+stateDiagram-v2
+direction LR
+    LogicalPlan --> OptimizedLogicalPlan: Optimize
+    OptimizedLogicalPlan --> PhysicalPlan:push down
+    note right of PhysicalPlan
+      Note: NestedOperator stays in PhysicalPlan\nafter push down for post-processing.
+    end note
+
+    state "Logical Plan Tree" as LogicalPlan
+    state LogicalPlan {
+        logState1: Project
+        logState2: Nested
+        logState3: ...
+
+        logState1 --> logState2
+        logState2 --> logState3
+        logState3 --> Relation
+    }
+
+    state "Optimized Logical Plan Tree" as OptimizedLogicalPlan
+    state OptimizedLogicalPlan {
+        optState1: Project
+        optState2: Nested
+
+        optState1 --> optState2
+        optState2 --> IndexScanBuilder
+    }
+
+    state "Physical Plan Tree" as PhysicalPlan
+    state PhysicalPlan {
+        phyState1: ProjectOperator
+        phyState2: NestedOperator
+
+        phyState1 --> phyState2
+        phyState2 --> IndexScan
+    }
+```
+
+### 2.2 Sequence Diagram for Nested SELECT Clause Query Push Down
+Nested function sequence diagram illustrating query execution from parsing to OpenSearch DSL push down.
+
+```mermaid
+sequenceDiagram
+    participant SQLService
+    participant ParserBaseRoot
+    participant AstExpressionBuilder
+
+    participant QueryService
+    participant Analyzer
+    participant NestedAnalyzer
+    participant Planner
+    participant TableScanPushDown
+    participant OpenSearchRequestBuilder
+    participant DefaultImplementor
+
+%% Parsing
+SQLService->>+ParserBaseRoot:visitRoot
+  ParserBaseRoot->>+AstExpressionBuilder:visitScalarFunction
+  AstExpressionBuilder-->>-ParserBaseRoot:Function
+ParserBaseRoot-->>-SQLService:UnresolvedPlan
+%% Analysis
+SQLService->>+QueryService:analyze
+  QueryService->>+Analyzer:visitProject
+    Analyzer->>+NestedAnalyzer:visitFunction
+    NestedAnalyzer-->>-Analyzer:LogicalNested
+  Analyzer-->>-QueryService:UnresolvedPlan
+    
+  %% planner optimization
+  QueryService->>+Planner:plan
+    Planner->>+TableScanPushDown:apply
+      TableScanPushDown->>+OpenSearchRequestBuilder:pushDownNested
+
+      Note over TableScanPushDown, OpenSearchRequestBuilder: returns false keeping<br>LogicalNested in plan tree
+
+      OpenSearchRequestBuilder-->>-TableScanPushDown:boolean
+    TableScanPushDown-->>-Planner:LogicalPlan
+    %% planner implementation
+    Planner->>+DefaultImplementor:visitNested
+    DefaultImplementor-->>-Planner:NestedOperator
+  Planner-->>-QueryService:PhysicalPlan
+QueryService-->>-SQLService:PhysicalPlan
+```
+
+### 2.3 Sequence Diagram for Nested SELECT Clause Post-processing
+Nested function sequence diagram illustrating the flattening of the OpenSearch response. Flattening the response from OpenSearch changes the nested types structure by making the full path of an object the key, and the object it refers to the value.
+
+**Sample input:**
```

Comment:
Could you include an example where we do a little more flattening?  Such as two nested objects with two fields each to show how we do the cartesian product?